### DISCUSSION TITLE
How do I determine if I'm at the far left or right edge of the screen in a kitten?

### DISCUSSION THREAD

**Discussion by mwpardue** [kovidgoyal/kitty#6003] [Q&A] [answered]
Hi! I'm interested in modifying the vim-tmux-navigator kitten so that I can use two keybinds to cycle left and right through windows, vim windows, and tabs. My thought is that when the current window is at either far side of the screen and you press the left or right keybinds it will move to the previous or next_tab, respectively. The problem I'm running into is that I'm not sure exactly how to tell when I'm at the farthest window to either side. 

I'm only really concerned with left and right, and I don't really ever use any layouts other than vertical, tall, and splits layouts, and I doubt I'd ever have more than two windows on either the top or bottom half of the screen. IDK if that makes the logic any easier, but maybe.

How do I tell when I'm at the far left or right edge of the window?

**Comment by kovidgoyal** [comment]
There are already actions to move to left/right/top/bottom neighboring
windows just use those.

**Reply by mwpardue** [reply]
> There are already actions to move to left/right/top/bottom neighboring windows just use those.

I'm sorry, I probably was not too clear, or maybe I'm misunderstanding your reply. 

Say I am using ctrl+l to move to the right neighboring window. When I reach the far right window and press ctrl+l again, I'd like it to move to the next_tab, if that tab exists. My problem is I don't know how to tell when it reaches that last window and needs to send next_tab instead. I would want to do the same thing with ctrl+h, but moving to neighboring_window left and once at the far left window move to previous tab, if it exists. 

I found another discussion where another user asked for a similar thing and they got it to work but never posted the code. Unfortunately I can't find the discussion now to link it.

**Comment by kovidgoyal** [comment]
It's layout dependent. Say in tall layout the last window will be the
tall window which unless you are mirroring will be the right most
window. By last here I mean last in the windows array of the kitty @ ls
output.

**Reply by mwpardue** [reply]
So there is not a way to find if I'm physically at the right-most or left-most border of the screen? I would need to rely solely on the numerical ID? I really doubt I'll use more than 2 split windows most of the time so I could probably use that method with little to no issue. I'd just have to assume window 1 is on the left half and window 2 on the right half, right?

**Reply by kovidgoyal** [reply]
Not id, *position* in the windows list.

**Reply by mwpardue** [reply]
> Not id, *position* in the windows list.

Oh, right. Reading is hard. 🙂 Thank you for your help and I'll post the final code back here when finished in case anybody in the future comes looking. 

Thanks!!

**Reply by mwpardue** [reply]
I was wrong in the first post here, I will not really use splits layout, rather grid and stacks will mostly work for my needs and is likely all the layouts I'll have enabled long-term. 

I mention that because I've been doing some testing in grid format to get a handle on how the 'kitty @ ls' array arranges the windows, and I'm not really seeing a good way to accomplish what I want based on the position in the list. Full disclaimer, though, I am not a programmer and am very likely overlooking an easy solution. I'm also assuming by position you meant the order it appears in the list because I don't see a 'position' property or anything analogous to that.

In my testing, I open four windows and as I open each I navigate to a directory called 'a', 'b', 'c', and 'd' respectively so I can easily track the windows in 'kitty @ ls'. I'm not doing anything in between opening, just 'cmd+enter', 'cd /testing/a', 'cmd+enter', 'cd /testing/b', rinse and repeat to four windows. Then I run 'kitty @ ls' and get the following, with window ID in parentheses:

|DIR(ID)|DIR(ID)|
|-------|--------|
|A(1)|C(3)|
|B(2)|D(4)|

Now let's say I close that Kitty instance and open a new one and this time open the same directories, but instead of opening each new window from the latest window (open B from A, C from B, D from C) I navigate back to the first window before opening the new window (open B from A, C from A, D from A). That gives me the following arrangement:

|DIR(ID)|DIR(ID)|
|-------|--------|
|A(1)|C(3)|
|D(4)|B(2)|

I'm not sure how to leverage the contents of 'kitty @ ls' to determine whether I'm at the left or right edge of the screen. What am I missing? Or how does the 'neighboring_window' command know when there is no neighboring window? Is there something returned by that function that I can use? 

Again, I apologize if I'm just missing something obvious, because I know there is some way to determine physical position on screen but I just can't quite see it.

**Comment by mwpardue** [comment]
I found the other discussion where this request was brought up but final code never posted. It was discussion #[4363](https://github.com/kovidgoyal/kitty/discussions/4363)

**Comment by page-down** [comment, marked as answer]
> ... but final code never posted. It was discussion 4363

No, in December 2021 I already said I had tried the so-called "final code".
And that is not what you want.

> ... use two keybinds to cycle ... through windows ...
> ... it will move to the previous or next_tab ...

```python
from typing import List

from kitty.boss import Boss
from kittens.tui.handler import result_handler


def main(args: List[str]) -> None:
    pass


@result_handler(no_ui=True)
def handle_result(args: List[str], answer: str, target_window_id: int, boss: Boss) -> None:
    prev = len(args) >  1
    tm = boss.active_tab_manager
    tab = boss.active_tab
    if tm is None or tab is None:
        return
    if prev:
        if tab.windows.active_group_idx == 0 and len(tm.tabs) > 1:
            boss.previous_tab()    
        else:    
            tab.previous_window()   
    else:    
        if tab.windows.active_group_idx >= len(tab.windows) - 1 and len(tm.tabs) > 1:
            boss.next_tab()    
        else:    
            tab.next_window()
```

```shell
kitty \
    -o 'map ctrl+shift+left kitten next_window_or_tab.py -1' \
    -o 'map ctrl+shift+right kitten next_window_or_tab.py'
```

The above code will first switch windows in the current tab, and will jump to the next tab when it reaches the first or last one.
It does not move to the first or last window in the tab after switching tabs.

UPDATE:

The following code will move to another tab on the left and right edges. e.g. if the current window uses a vertical layout, pressing the right arrow key combination will move to the next tab (usually the next tab is on the right).

```python
from typing import List

from kitty.boss import Boss
from kittens.tui.handler import result_handler


def main(args: List[str]) -> None:
    pass


@result_handler(no_ui=True)
def handle_result(args: List[str], answer: str, target_window_id: int, boss: Boss) -> None:
    if len(args) != 2:
        return
    direction = args[1]
    if direction not in ('left', 'right'):
        return

    tm = boss.active_tab_manager
    tab = boss.active_tab
    if tm is None or tab is None:
        return

    neighbor = tab.neighboring_group_id(direction)
    if neighbor is not None:
        tab.windows.set_active_group(neighbor)
    elif len(tm.tabs) > 1:
        if direction == 'left':
            boss.previous_tab()
        else:
            boss.next_tab()
```

```shell
kitty \
    -o 'map ctrl+shift+left kitten go_left_or_right.py left' \
    -o 'map ctrl+shift+right kitten go_left_or_right.py right'
```

UPDATE2:
The following code will switch the window from that direction after switching tabs, e.g. when going right, switch to the leftmost window in the next tab.

```python
from typing import List

from kitty.boss import Boss
from kitty.tabs import Tab
from kittens.tui.handler import result_handler


def main(args: List[str]) -> None:
    pass


@result_handler(no_ui=True)
def handle_result(args: List[str], answer: str, target_window_id: int, boss: Boss) -> None:
    if len(args) != 2:
        return
    direction = args[1]
    if direction not in ('left', 'right'):
        return

    tm = boss.active_tab_manager
    tab = boss.active_tab
    if tm is None or tab is None:
        return

    def neighboring_window(t: Tab, which: str) -> bool:
        neighbor = t.neighboring_group_id(which)
        if neighbor is not None:
            t.windows.set_active_group(neighbor)
            return True
        return False

    if neighboring_window(tab, direction):
        return
    elif len(tm.tabs) > 1:
        if direction == 'left':
            boss.previous_tab()
        else:
            boss.next_tab()

        tab = boss.active_tab
        if tab is not None and len(tab.windows) > 1:
            direction = 'right' if direction == 'left' else 'left'
            while neighboring_window(tab, direction):
                continue
```

**Reply by mwpardue** [reply]
Sure, but the very last reply by OP said they made it work but they never posted it. Their ultimate goal was the same as mine. Sorry if I made it sound like I thought that code posted should work. 

That being said, can you help me with figuring out how to determine when I'm at the far edge of the screen? I've been looking at kitty's code a lot but I am no programmer and can't figure out exactly how neighboring_window works. I'm decent enough at programming, though, that I think if somebody could help me with that bit then I could get the rest of the way. At least I hope I can. 

>The above code does not move to the first or last window in the tab after switching tabs.

I'm sorry, I reread your reply and this isn't precisely what I'm after. Not sure if I'm misreading it, but what I'm after is to execute the "neighboring_window right" action when I press this keybind for the right direction if there is a neighboring window in that direction, if I'm at the far right edge of the screen already then execute "next_tab" instead. Same if I'm moving left, "neighboring_window left" if there is a neighboring window, else "previous_tab" if there's not. Of course that last action would assume there is an available tab, but I believe I have that part worked out it's just the neighboring window detection that I'm not sure of.

**Reply by page-down** [reply]
> ... ultimate goal was the same ...

I'm not so sure about this, but it doesn't matter anymore.

> ... can you help me with figuring out ...

For this reason I have written the code for you, the one above.
What I mean by that last line is that after switching tabs no more switching window is done, you can add it if you need.

If you have any further questions feel free to ask.

**Reply by page-down** [reply]
@mwpardue
I have updated the reply above.
Another example has been added.

If the vertical layout is being used, then pressing the right arrow key will take you to the next tab.
So shortcuts will not switch between windows in the same vertical layout.

**Reply by mwpardue** [reply]
@page-down 
Thanks for writing that out! I really appreciate it, and I see where I was going wrong now, I never tried using neighboring_group_id, I kept focusing on neighboring_window or just neighbor because I thought neighboring_group_id was something unrelated to what I was doing. What you wrote covers the request, for sure.

One last question. Currently on moving into a tab using next/previous_tab it activates the most recently active window. Is it possible to activate the first window from that direction instead? 

Example, if needed: Assuming I have two windows open in tab 1 in any layout besides vertical, and use the go_left_or_right.py script to move left from tab 2 to tab 1 it currently defaults to the last active window in that tab. I would like to see it always activate the farthest right window instead. If I'd used the script to move right instead, assuming only 2 tabs total, then I'd like to see it activate the farthest left window in tab 1. 

Given the one-binding approach I'm trying to take to movements in each direction this seems to make more sense. It would be amazing if it could stay in the same row, too, but that may be more difficult than it's worth. I honestly don't know.

Again, I really appreciate your help with the previous script, it does precisely what I asked for. I'm not really asking you to write this particular change for me, if you're able to point me in the right direction that would be great! It's the only way I'll learn. If it's not a reasonable thing to do then no big deal.

**Reply by page-down** [reply]
> ... Is it possible to activate the first window from that direction instead?

If you don't want to learn the window structure of the layout, just keep switching windows in the opposite direction after switching tabs until there is none.
I have updated my reply to include the new code.

> ... It would be amazing if it could stay in the same row, too ...

You need to understand the structure of the window in the layout, learn the relevant code and use the appropriate algorithm (which will not be too complicated) to get the respective group id.
I will leave this part to you.

You can click to accept the answer to mark it as solved.
Maybe this will help other people as well.

**Reply by mwpardue** [reply]
> > ... Is it possible to activate the first window from that direction instead?
> 
> If you don't want to learn the window structure of the layout, just keep switching windows in the opposite direction after switching tabs until there is none. I have updated my reply to include the new code.

I honestly didn't consider that. Since I'm only going to be using the grid layout (and probably stacks but that's less relevant) then it would probably be pretty easy to achieve what I'm looking for, and worst case I can do as you said and move back the opposite direction to the far edge. I'd rather come up with a real solution, though. You've given me a place to start and I really appreciate the help you've provided. 

> > ... It would be amazing if it could stay in the same row, too ...
> 
> You need to understand the structure of the window in the layout, learn the relevant code and use the appropriate algorithm (which will not be too complicated) to get the respective group id. I will leave this part to you.

As I said before, I think with what you've already provided I can probably make this work. I think I was just looking for somebody with more experience to provide some guidance or feedback, and you've done that. Thanks again!

> You can click to accept the answer to mark it as solved. Maybe this will help other people as well.

While your previous reply was exactly what I asked for and will be marked as the answer, my ultimate goal was to modify the vim-kitty-navigator kitten with this behavior, so I want to leave this link to the discussion over there just in case anybody else is searching for this specifically. It has the modifications I made and will be updated if/when I can make further changes. [vim-kitty-navigator discussion](https://github.com/knubie/vim-kitty-navigator/issues/39)

@page-down Thanks again!!
